### PR TITLE
Pyroscope: Change the data source id with migrations

### PR DESCRIPTION
**Pull request by aocenas**
After [renaming Phlare data source to Grafana Pyroscope](https://github.com/grafana/grafana/pull/66989) the ID was still `phlare`. This change the ID and adds migrations to change dashboards and data sources in DB to the new ID.

To test:

- While on main branch, create Grafana Pyroscope datasource and a dashboard with Flamegraph panel with query from that data source
- Checkout this branch and restart
- Data source and dashboard should still work without any errors

# Release notice breaking change

Due to renaming of the Phlare data source to Grafana Pyroscope also the type ID changed from `phlare` to `grafana-pyroscope`. This affects any provisioning of the data source, where the property `type: phlare` has to be changed to `type: grafana-pyroscope`. This also affects alerts, recorded queries and query history which will need to be recreated.

Dashboard panels and any instances of the old Phlare data source should be automatically migrated and don't require any user action.

### CODE REVIEW

**Review by marefr — COMMENTED** [review]
Should work, but think you might need to handle Grafana Alerting/Recorded Queries, e.g. queries stored in the database that are executing in Grafana backend. However I guess this datasource is not applicable for these features or is it?

**Comment by aocenas** [comment]
@marefr yeah I don't really see alerting on this. I would say even if somebody does that in some way it's probably low impact. But thanks we should mention it in the breaking changes if this gets merged.

**Comment by aocenas** [comment]
@marefr btw are we doing down migrations? I don't see anything anywhere but I thought we do that somehow.

**Comment by marefr** [comment]
Came to remember https://github.com/grafana/grafana/pull/25760 and https://github.com/grafana/grafana/pull/25807. Doesn't look like migrations was considered in that one though. Part of not renaming the plugin id was backward and forward compatibility concerns it seems.

**Comment by aocenas** [comment]
@marefr yeah we also discussed not changing but considering that Phalre didn't have a stable release yet and could be considered experimental still, we assume the breaking changes won't affect many people.

**Comment by aocenas** [comment]
@marefr can you do a review of this on the technical level, whether the migrations make sense. I didn't do any migration for a long time (and none for the dashboards) so don't want to cargocult something that should not be used anymore and nobody in our squad have experience with this either.

File: pkg/services/sqlstore/migrations/phlare_to_grafana_pyroscope_rename.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package migrations
+
+import "github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+
+func changePhlareIdToGrafanaPyroscope(mg *migrator.Migrator) {
+	mg.AddMigration("update datasource type",
```
Comment (ryantxu) [inline pkg/services/sqlstore/migrations/phlare_to_grafana_pyroscope_rename.go:6]:
```suggestion
	mg.AddMigration("migrate phlare data_source to pyroscope",
```
the name here is used as a key, and must be unique across all migrations, so lets pick something more specific

**Review by ryantxu — APPROVED** [review]

**Comment by bergquist** [comment]
Note that this change is breaking so rolling back will break Phlare datasources configured when using 9.5x. We have been forced to rollback some instances during the previous releases so the chances of rolling back 10.0 is high. 

What you could do instead is a soft migration in memory when reading the value from the database. Basically an if statement in the datasource config get/list path that replaces phlare with pyroscope. Once 10.1. is released you could do the database migration.

**Comment by aocenas** [comment]
@bergquist Yeah I started this thinking we have down migrations, without them this is more problematic I agree.

> How important is it that we rename this id in Grafana 10?

We wanted to do it now mainly because we think that any breaking change will have much less impact than later when we may have stable Grafana Pyroscope backend and more users will start to use it.

I am thinking about how to easily implement what you are suggesting. I think we could just leave Phlare data source in Grafana, in a way that both Phlare and Grafana Pyroscope will point to the same code and adding some attribute saying Phlare is hidden from the Create data source page. That would effectively create an alias for the ID and I think that could be less invasive and error prone. Somewhere in the future we can then do this migration and even without down migration, it won't create a problem on rollback and maybe add some warning for the provisioning before that.

File: pkg/services/sqlstore/migrations/phlare_to_grafana_pyroscope_rename.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+		migrator.NewRawSQLMigration("").
+			SQLite(`UPDATE data_source SET type='grafana-pyroscope' WHERE type='phlare';`).
+			Postgres(`UPDATE data_source SET type='grafana-pyroscope' WHERE type='phlare';`).
+			Mysql(`UPDATE data_source SET type='grafana-pyroscope' WHERE type='phlare';`),
```
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/services/sqlstore/migrations/phlare_to_grafana_pyroscope_rename.go:10]: Agree

**Review by marefr — COMMENTED** [review]
> @marefr can you do a review of this on the technical level, whether the migrations make sense. I didn't do any migration for a long time (and none for the dashboards) so don't want to cargocult something that should not be used anymore and nobody in our squad have experience with this either.

The database migration looks good with Ryan's suggestion. Dashboard migration there's probably others better suited to review. 

> I am thinking about how to easily implement what you are suggesting. I think we could just leave Phlare data source in Grafana, in a way that both Phlare and Grafana Pyroscope will point to the same code and adding some attribute saying Phlare is hidden from the Create data source page. That would effectively create an alias for the ID and I think that could be less invasive and error prone. Somewhere in the future we can then do this migration and even without down migration, it won't create a problem on rollback and maybe add some warning for the provisioning before that.

Doesn't feel optimal, but maybe a way forward. Probably need to see it before I fully understand what this solution implies. In general, feels like Grafana plugin management should be able to provide support for renaming of a plugin somehow, creating an alias or similar. Maybe even without database migration to allow backward and forward compatibility? It has come up before, without really looking into what such solution might mean.

**Comment by aocenas** [comment]
@marefr @bergquist I created PR with the approach I mentioned: https://github.com/grafana/grafana/pull/67847. It is more code but at the same time, it seems to me like this is less error-prone and easier to understand than migrations or doing some translation in the read path somewhere.

**Pull request closed without merging**